contracts/src/periphery/ControllerV1.sol

```diff
@@ -101,36 +101,83 @@ contract ControllerV1 is Ownable {
         if (asset0 > asset1) (asset0, asset1) = (asset1, asset0);
-        StoreKey key = PoolConfigStoreLib.keyFromAssetsUnchecked(asset0, asset1);
-        pools[key] = Pool(asset0, asset1);
+        StoreKey key = StoreKeyLib.keyFromAssetsUnchecked(asset0, asset1);
+
+        uint256 maybe_index = _poolIndices[key];
+        if (maybe_index == 0) {
+            _pools.push(Pool(asset0, asset1));
+            maybe_index = _pools.length;
+            _poolIndices[key] = maybe_index;
+        }
 
         emit PoolConfigured(asset0, asset1, tickSpacing, bundleFee, unlockedFee);
         ANGSTROM.configurePool(asset0, asset1, tickSpacing, bundleFee, unlockedFee);
     }
 
     function removePool(address asset0, address asset1) external {
-        _checkOwner();
+        _checkOwnerOrNode();
 
         if (asset0 > asset1) (asset0, asset1) = (asset1, asset0);
-        StoreKey key = PoolConfigStoreLib.keyFromAssetsUnchecked(asset0, asset1);
-        PoolConfigStore configStore = ANGSTROM.configStore();
-        uint256 poolIndex = 0;
-
-        uint256 totalEntries = configStore.totalEntries();
-        ConfigEntry entry;
-        while (true) {
-            entry = configStore.getWithDefaultEmpty(key, poolIndex);
-            if (!entry.isEmpty()) break;
-            poolIndex++;
-            if (poolIndex >= totalEntries) {
-                revert NonexistentPool(asset0, asset1);
+        StoreKey key = StoreKeyLib.keyFromAssetsUnchecked(asset0, asset1);
+
+        unchecked {
+            uint256 index_plus_one = _poolIndices[key];
+            if (index_plus_one == 0) revert NonexistentPool(asset0, asset1);
+            uint256 index = index_plus_one - 1;
+
+            uint256 length = _pools.length;
+            if (index_plus_one < length) {
+                Pool memory last_pool = _pools[length - 1];
+                StoreKey last_key =
+                    StoreKeyLib.keyFromAssetsUnchecked(last_pool.asset0, last_pool.asset1);
+                _pools[index] = last_pool;
+                _poolIndices[last_key] = index_plus_one;
             }
+
+            _poolIndices[key] = 0;
+            _pools.pop();
+
+            PoolConfigStore config_store = ANGSTROM.configStore();
+            (int24 tick_spacing, uint24 bundle_fee) = config_store.get(key, index);
+            emit PoolRemoved(asset0, asset1, tick_spacing, bundle_fee);
+            ANGSTROM.removePool(key, config_store, index);
         }
+    }
 
-        emit PoolRemoved(asset0, asset1, entry.tickSpacing(), entry.bundleFee());
-        ANGSTROM.removePool(key, configStore, poolIndex);
+    struct PoolUpdate {
+        address assetA;
+        address assetB;
+        uint24 bundleFee;
+        uint24 unlockedFee;
+    }
+
+    function batchUpdatePools(PoolUpdate[] calldata updates) external {
+        _checkOwnerOrNode();
+
+        ConfigEntryUpdate[] memory entry_updates = new ConfigEntryUpdate[](updates.length);
+
```

Comment:
Talked to Karthik, Can we add a check here to assert that both the bundleFee and unlockedFee are at Max 1% fee